stuffer/s2n_stuffer_text.c

```diff
@@ -206,3 +207,80 @@ int s2n_stuffer_init_ro_from_string(struct s2n_stuffer *stuffer, uint8_t *data,
 
     return S2N_SUCCESS;
 }
+
+/* If we call va_start or va_copy there MUST be a matching call to va_end,
+ * so we should use DEFER_CLEANUP with our va_lists.
+ * Unfortunately, some environments implement va_list in unexpected ways
+ * (for example, as an array).
+ * To avoid any surprises, just wrap the va_list in our own struct.
+ */
+struct s2n_va_list {
+    va_list va_list;
+};
+
+static void s2n_va_list_cleanup(struct s2n_va_list *list)
+{
+    if (list) {
+        va_end(list->va_list);
+    }
+}
+
+int s2n_stuffer_vnprintf(struct s2n_stuffer *stuffer, uint32_t max_size,
+        const char *format, va_list vargs_in)
+{
+    POSIX_PRECONDITION(s2n_stuffer_validate(stuffer));
+    POSIX_ENSURE_REF(format);
+
+    /* vsnprintf consumes the va_list, so copy it first */
+    DEFER_CLEANUP(struct s2n_va_list vargs_1 = { 0 }, s2n_va_list_cleanup);
+    va_copy(vargs_1.va_list, vargs_in);
+
+    /* The first call to vsnprintf calculates the size of the formatted string.
+     * str_len does not include the one byte vsnprintf requires for a trailing '\0',
+     * so we need one more byte.
+     */
+    int str_len = vsnprintf(NULL, 0, format, vargs_1.va_list);
+    POSIX_ENSURE_GTE(str_len, 0);
+    POSIX_ENSURE_LT((unsigned int) str_len, UINT32_MAX);
+    uint32_t mem_size = MIN((uint32_t) str_len + 1, max_size);
+
+    bool previously_tainted = stuffer->tainted;
+    char *str = s2n_stuffer_raw_write(stuffer, mem_size);
+    stuffer->tainted = previously_tainted;
+    POSIX_GUARD_PTR(str);
+
+    /* vsnprintf again consumes the va_list, so copy it first */
+    DEFER_CLEANUP(struct s2n_va_list vargs_2 = { 0 }, s2n_va_list_cleanup);
+    va_copy(vargs_2.va_list, vargs_in);
+
+    /* This time, vsnprintf actually writes the formatted string */
+    int written = vsnprintf(str, mem_size, format, vargs_2.va_list);
+    POSIX_ENSURE_GTE(written, 0);
+
+    /* We don't actually use c-strings, so erase the final '\0' */
+    POSIX_GUARD(s2n_stuffer_wipe_n(stuffer, 1));
```

Comment:
`\0` isn't necessary for writing to stdout. `\0` is only necessary for methods that deal with c-strings. It's how you indicate the end / length of the string without needing to pass around a separate length variable. But c-strings are notorious for causing memory / security issues, because relying on the existence of `\0` is risky. We're using a stuffer, which tracks the length for us. We don't need '\0'. We have a much more reliable length variable.

If you're thinking "but can you use a non-c-string with printf?!" the answer is yes: `fprintf(stdout, "%.*s", size, str)`.